In [1]:
import time
import datetime
principio = datetime.datetime.now()
principio

datetime.datetime(2024, 1, 10, 20, 47, 56, 554807)

In [2]:
# Packages
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sktime.utils.plotting import plot_series
import seaborn as sns
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.difference import Differencer
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.arima import AutoARIMA
from sktime.performance_metrics.forecasting import mean_absolute_percentage_error
from sktime.performance_metrics.forecasting import MeanSquaredError
from sklearn.metrics import r2_score



# Read Data
data = pd.read_csv('../data/datos sin días.csv')
data.set_index('Fecha', inplace=True)
transformer = LogTransformer()
data.index = pd.to_datetime(data.index)
log_data = transformer.fit_transform(data)


Bad key savefig.frameon in file C:\Users\manue\anaconda3\envs\TecnicasPrediccion\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 421 ('savefig.frameon : True')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.4.3/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.level in file C:\Users\manue\anaconda3\envs\TecnicasPrediccion\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 472 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.4.3/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.fileo in file C:\Users\manue\anaconda3\envs\TecnicasPrediccion\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 473 ('verbose.fileo  : sys.stdout  # a log filename, sys.std

In [3]:
transf_diff=Differencer(lags=[1])

In [4]:
y_train, y_test = temporal_train_test_split(y =data.astype('float64').to_period('d'), test_size=31)
log_y_train, log_y_test = temporal_train_test_split(y = log_data.astype('float64').to_period('d'), test_size=31)


In [5]:
fh = np.arange(len(y_test)) + 1

In [7]:
import warnings
warnings.filterwarnings("ignore")

from itertools import product
error_action="ignore"

seasonal = [True, False]
stationary = [True, False]
information_criterion = ["aic", "bic", "hqic", "oob"]
scoring = ['mse', 'mae']
sp_values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 30, 31]

param_combinations = list(product(seasonal, stationary, information_criterion, scoring, sp_values))

for params in param_combinations:
    forecaster = AutoARIMA(
        seasonal=params[0],
        stationary=params[1],
        information_criterion=params[2],
        scoring=params[3],
        sp=params[4],
        suppress_warnings=True)
    forecaster.fit(log_y_train)
    log_y_pred = forecaster.predict(fh)
    y_pred=np.exp(log_y_pred)
    mape = mean_absolute_percentage_error(y_test[0:31], y_pred[0:31])
    print(f'con seasonal {params[0]}, stationary {params[1]}, ic {params[2]}, scoring {params[3]}, sp {params[4]} MAPE: {mape}')

con seasonal True, stationary True, ic aic, scoring mse, sp 1 MAPE: 0.5008937876516164
con seasonal True, stationary True, ic aic, scoring mse, sp 2 MAPE: 0.696107373726597
con seasonal True, stationary True, ic aic, scoring mse, sp 3 MAPE: 0.687546683984665
con seasonal True, stationary True, ic aic, scoring mse, sp 4 MAPE: 0.6735095095321186
con seasonal True, stationary True, ic aic, scoring mse, sp 5 MAPE: 0.5109270809093985
con seasonal True, stationary True, ic aic, scoring mse, sp 6 MAPE: 0.4960766473743328
con seasonal True, stationary True, ic aic, scoring mse, sp 7 MAPE: 0.3051574107940964
con seasonal True, stationary True, ic aic, scoring mse, sp 8 MAPE: 0.5137605933878617
con seasonal True, stationary True, ic aic, scoring mse, sp 9 MAPE: 0.49294213739910625
con seasonal True, stationary True, ic aic, scoring mse, sp 10 MAPE: 0.6447592598295754
con seasonal True, stationary True, ic aic, scoring mse, sp 11 MAPE: 0.4568692462410879
con seasonal True, stationary True, ic aic

con seasonal True, stationary True, ic oob, scoring mse, sp 11 MAPE: 0.4568692462410879
con seasonal True, stationary True, ic oob, scoring mse, sp 12 MAPE: 0.5151840944599273
con seasonal True, stationary True, ic oob, scoring mse, sp 30 MAPE: 0.693602596600012
con seasonal True, stationary True, ic oob, scoring mse, sp 31 MAPE: 0.6907699701223925
con seasonal True, stationary True, ic oob, scoring mae, sp 1 MAPE: 0.5008937876516164
con seasonal True, stationary True, ic oob, scoring mae, sp 2 MAPE: 0.696107373726597
con seasonal True, stationary True, ic oob, scoring mae, sp 3 MAPE: 0.687546683984665
con seasonal True, stationary True, ic oob, scoring mae, sp 4 MAPE: 0.6735095095321186
con seasonal True, stationary True, ic oob, scoring mae, sp 5 MAPE: 0.5109270809093985
con seasonal True, stationary True, ic oob, scoring mae, sp 6 MAPE: 0.4960766473743328
con seasonal True, stationary True, ic oob, scoring mae, sp 7 MAPE: 0.3051574107940964
con seasonal True, stationary True, ic oob

con seasonal True, stationary False, ic hqic, scoring mae, sp 6 MAPE: 0.4302036691242656
con seasonal True, stationary False, ic hqic, scoring mae, sp 7 MAPE: 0.3270840313072271
con seasonal True, stationary False, ic hqic, scoring mae, sp 8 MAPE: 0.42704603151977555
con seasonal True, stationary False, ic hqic, scoring mae, sp 9 MAPE: 0.6305284003630207
con seasonal True, stationary False, ic hqic, scoring mae, sp 10 MAPE: 0.3978672944131838
con seasonal True, stationary False, ic hqic, scoring mae, sp 11 MAPE: 0.4722863027036086
con seasonal True, stationary False, ic hqic, scoring mae, sp 12 MAPE: 0.44224850116126363
con seasonal True, stationary False, ic hqic, scoring mae, sp 30 MAPE: 0.41090375373150645
con seasonal True, stationary False, ic hqic, scoring mae, sp 31 MAPE: 0.48670177317707225
con seasonal True, stationary False, ic oob, scoring mse, sp 1 MAPE: 0.42704603151977555
con seasonal True, stationary False, ic oob, scoring mse, sp 2 MAPE: 0.5145831898012632
con seasonal 

con seasonal False, stationary True, ic hqic, scoring mse, sp 1 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 2 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 3 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 4 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 5 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 6 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 7 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 8 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 9 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 10 MAPE: 0.5008937876516164
con seasonal False, stationary True, ic hqic, scoring mse, sp 11 MAPE: 0.5008937876516164
con seasonal False,

con seasonal False, stationary False, ic bic, scoring mse, sp 9 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mse, sp 10 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mse, sp 11 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mse, sp 12 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mse, sp 30 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mse, sp 31 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mae, sp 1 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mae, sp 2 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mae, sp 3 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mae, sp 4 MAPE: 0.4906409449393137
con seasonal False, stationary False, ic bic, scoring mae, sp 5 MAPE: 0.4906409449393137
con seasonal Fal

con seasonal True, stationary True, ic aic, scoring mse, MAPE: 0.3051574107940964

con seasonal True, stationary True, ic aic, scoring mae, MAPE: 0.3051574107940964

con seasonal True, stationary True, ic hqic, scoring mse, MAPE: 0.3051574107940964

con seasonal True, stationary True, ic hqic, scoring mae, MAPE: 0.3051574107940964

con seasonal True, stationary True, ic oob, scoring mae, MAPE: 0.3051574107940964


In [8]:
forecaster = AutoARIMA(
    seasonal=True,
    stationary=True,
    information_criterion="aic",
    scoring="mse",
    sp=7,
    suppress_warnings=True)
forecaster.fit(log_y_train)

AutoARIMA(sp=7, stationary=True, suppress_warnings=True)

In [9]:
log_y_pred = forecaster.predict(fh)

In [10]:
y_pred=np.exp(log_y_pred)

In [11]:
mape = mean_absolute_percentage_error(y_test[0:31], y_pred[0:31])
mse = MeanSquaredError()
mse = mse(y_test[0:31], y_pred[0:31])
r2 = r2_score(y_test, y_pred)

print(f'MAPE: {mape}')
print(f'MSE: {mse}')
print(f'R2 Score: {r2}')

MAPE: 0.3051574107940964
MSE: 26153315.84357026
R2 Score: 0.6057691106557495


Ahora sin log

In [12]:
forecaster = AutoARIMA(sp=7,suppress_warnings=True)
forecaster.fit(y_train)

AutoARIMA(sp=7, suppress_warnings=True)

In [13]:
y_pred = forecaster.predict(fh)

In [14]:
mape = mean_absolute_percentage_error(y_test[0:31], y_pred[0:31])
mse = MeanSquaredError()
mse = mse(y_test[0:31], y_pred[0:31])
r2 = r2_score(y_test, y_pred)

print(f'MAPE: {mape}')
print(f'MSE: {mse}')
print(f'R2 Score: {r2}')

MAPE: 0.36443183478784985
MSE: 33118315.14727451
R2 Score: 0.5007798279886926


In [15]:
sp_values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 30, 31]
for i in sp_values:
    forecaster = AutoARIMA(sp=i, suppress_warnings=True)
    forecaster.fit(y_train)
    y_pred = forecaster.predict(fh)
    
    mape = mean_absolute_percentage_error(y_test[0:31], y_pred[0:31])
    
    mse = MeanSquaredError()
    mse_value = mse(y_test[0:31], y_pred[0:31])
    
    r2 = r2_score(y_test[0:31], y_pred[0:31])

    print(f'MAPE con sp={i}: {mape}')
    print(f'MSE con sp={i}: {mse_value}')
    print(f'R2 Score con sp={i}: {r2}')

MAPE con sp=1: 0.49978361300548485
MSE con sp=1: 51647567.81421771
R2 Score con sp=1: 0.2214728444511075
MAPE con sp=2: 0.488380173899177
MSE con sp=2: 53774468.64376708
R2 Score con sp=2: 0.18941228239482755
MAPE con sp=3: 0.5533183316939516
MSE con sp=3: 57926259.52130554
R2 Score con sp=3: 0.1268288524460882
MAPE con sp=4: 0.48688114716210157
MSE con sp=4: 52606456.43291552
R2 Score con sp=4: 0.20701871117055604
MAPE con sp=5: 0.5051259846536055
MSE con sp=5: 52738475.54563647
R2 Score con sp=5: 0.20502867623465282
MAPE con sp=6: 0.5030398834790497
MSE con sp=6: 51315440.832997635
R2 Score con sp=6: 0.2264792733094887
MAPE con sp=7: 0.36443183478784985
MSE con sp=7: 33118315.14727451
R2 Score con sp=7: 0.5007798279886926
MAPE con sp=8: 0.47602600352021834
MSE con sp=8: 48071979.62306472
R2 Score con sp=8: 0.2753706874993186
MAPE con sp=9: 0.49978361300548485
MSE con sp=9: 51647567.81421771
R2 Score con sp=9: 0.2214728444511075
MAPE con sp=10: 0.5058569988360984
MSE con sp=10: 525721

In [16]:
final = datetime.datetime.now()
duracion = final-principio
print(principio)
print(final)
print(duracion)

2024-01-10 20:47:56.554807
2024-01-12 02:10:09.470828
1 day, 5:22:12.916021
